### PR TITLE
issue #3 - text alignment

### PR DESCRIPTION
**Pull request by wolski0420**
Third issue done!

### CODE REVIEW

File: text_align.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+def align_text(image: np.ndarray) -> np.ndarray:
+    """
+        Rotate sheet to get aligned, parallel text
+
+        :param image:
+            loaded image as numpy.ndarray type
+
+        :return:
+            rotated image as numpy.ndarray type
```
Comment (j-adamczyk) [inline text_align.py:18]: Let's use Numpydoc docstring style, see e.g. https://stackoverflow.com/questions/3898572/what-is-the-standard-python-docstring-format.

File: text_align.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
+from skimage.io import imread, imsave
+from skimage.filters import gaussian, threshold_otsu
+from skimage.color import rgb2gray, rgba2rgb
+from skimage.feature import canny
+from skimage.transform import probabilistic_hough_line, rotate
+from skimage import img_as_ubyte
+import numpy as np
```
Comment (j-adamczyk) [inline text_align.py:7]: Imports (libraries names) be ordered lexicographically

File: text_align.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            rotated image as numpy.ndarray type
+    """
+
+    # checking if image is gray, rgba or rgb
+    draft = image
+    if len(draft.shape) > 2:
+        if draft.shape[2] == 4:         # converting from rgba to rgb
+            draft = rgba2rgb(draft)
+        if draft.shape[2] == 3:         # converting from rgb to gray
+            draft = rgb2gray(draft)
```
Comment (j-adamczyk) [inline text_align.py:27]: I think we can assume that image is RGB and write this in the docstring

File: text_align.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+    # Otsu's method - returns threshold to determine text block
+    thresh: float = threshold_otsu(draft)
+
+    # creating binary image where (True = text) and converting it to gaussian blur
```
Comment (j-adamczyk) [inline text_align.py:32]: You can blur an image, you can't "convert" to gaussian blur

File: text_align.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    binary_img: np.ndarray = draft > thresh
+    gaussian_blur: np.ndarray = gaussian(binary_img, 3)
```
Comment (j-adamczyk) [inline text_align.py:34]: Consider bilateral filtering, it preserves lines better

File: text_align.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+    # detecting edges and adopting probabilistic line Hough transform
```
Comment (j-adamczyk) [inline text_align.py:36]: "adopting" sounds weird here, maybe "applying" would be better?

File: text_align.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    edges: np.ndarray = canny(gaussian_blur)
+    lines = probabilistic_hough_line(edges)
+
+    # every hough line needs to be transformed to slope value and then to the angle (like in linear functions)
+    slopes = [(y2 - y1) / (x2 - x1) if (x2 - x1) else 0 for (x1, y1), (x2, y2) in lines]
+    deg_angles = [np.degrees(np.arctan(x)) for x in slopes]
+
+    # getting most common degree value
```
Comment (j-adamczyk) [inline text_align.py:44]: Should be "the most common"

File: text_align.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    binary_img: np.ndarray = draft > thresh
+    gaussian_blur: np.ndarray = gaussian(binary_img, 3)
+
+    # detecting edges and adopting probabilistic line Hough transform
+    edges: np.ndarray = canny(gaussian_blur)
+    lines = probabilistic_hough_line(edges)
+
+    # every hough line needs to be transformed to slope value and then to the angle (like in linear functions)
+    slopes = [(y2 - y1) / (x2 - x1) if (x2 - x1) else 0 for (x1, y1), (x2, y2) in lines]
+    deg_angles = [np.degrees(np.arctan(x)) for x in slopes]
```
Comment (j-adamczyk) [inline text_align.py:42]: May be faster using Numpy broadcasting:
```
slopes = np.array([ ... ])
deg_angles = np.degrees(np.arctan(slopes))
```

File: text_align.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+    # getting most common degree value
+    histogram = np.histogram(deg_angles, bins=180)
+    rotation_angle: float = histogram[1][np.argmax(histogram[0])]
```
Comment (j-adamczyk) [inline text_align.py:46]: This magic indices should be commented

File: text_align.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    # rotating and returning image
+    img = rotate(image, angle=rotation_angle)
+    return img
+
+
+# example
+file_name: str = "skewed.jpg"
+img_file: np.ndarray = imread(file_name)
+rotated_image = align_text(img_file)
+imsave("rotated_" + file_name, img_as_ubyte(rotated_image))
```
Comment (j-adamczyk) [inline text_align.py:63]: Should be removed from final commited code or moved to the separate test file

**Comment by wolski0420** [comment]
I made changes as you suggested

**Review by j-adamczyk — APPROVED** [review]

**Review by j-adamczyk — APPROVED** [review]

**Review by konrrad — APPROVED** [review]

**Pull request merged**